File: src/test/java/cube8540/oauth/authentication/users/application/DefaultUserCredentialsServiceTest.java
package cube8540.oauth.authentication.users.application;

import cube8540.oauth.authentication.users.domain.User;
import cube8540.oauth.authentication.users.domain.UserCredentialsKeyGenerator;
import cube8540.oauth.authentication.users.domain.UserRepository;
import cube8540.oauth.authentication.users.domain.exception.UserNotFoundException;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.mockito.InOrder;

import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.RAW_CREDENTIALS_KEY;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.RAW_USERNAME;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.USERNAME;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.makeDefaultUser;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.makeEmptyUserRepository;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.makeKeyGenerator;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.makeUserRepository;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.inOrder;
import static org.mockito.Mockito.times;

@DisplayName("기본 유저 인증 서비스 테스트")
class DefaultUserCredentialsServiceTest {

    @Test
    @DisplayName("저장소에 저장 되지 않은 유저에게 인증키 할당")
    void grantCredentialsKeyToNotRegisteredUserInRepository() {
        UserRepository repository = makeEmptyUserRepository();

        UserCredentialsService service = new DefaultUserCredentialsService(repository);

        assertThrows(UserNotFoundException.class, () -> service.grantCredentialsKey(RAW_USERNAME));
    }

    @Test
    @DisplayName("저장소에 저장된 유저에게 인증키 할당")
    void grantCredentialsKeyToRegisteredUserInRepository() {
        User user = makeDefaultUser();
        UserRepository repository = makeUserRepository(USERNAME, user);
        UserCredentialsKeyGenerator keyGenerator = makeKeyGenerator();

        DefaultUserCredentialsService service = new DefaultUserCredentialsService(repository);
        InOrder inOrder = inOrder(user, repository);
        service.setKeyGenerator(keyGenerator);
        service.grantCredentialsKey(RAW_USERNAME);

        inOrder.verify(user).generateCredentialsKey(keyGenerator);
        inOrder.verify(repository, times(1)).save(user);
    }

    @Test
    @DisplayName("저장 되지 않은 유저 계정 인증")
    void accountAuthenticationToNotRegisteredUserInRepository() {
        UserRepository repository = makeEmptyUserRepository();

        DefaultUserCredentialsService service = new DefaultUserCredentialsService(repository);

        assertThrows(UserNotFoundException.class, () -> service.accountCredentials(RAW_USERNAME, RAW_CREDENTIALS_KEY));
    }

    @Test
    @DisplayName("유저 계정 인증")
    void accountAuthenticationToUser() {
        User user = makeDefaultUser();
        UserRepository repository = makeUserRepository(USERNAME, user);

        DefaultUserCredentialsService service = new DefaultUserCredentialsService(repository);
        InOrder inOrder = inOrder(user, repository);

        service.accountCredentials(RAW_USERNAME, RAW_CREDENTIALS_KEY);
        inOrder.verify(user, times(1)).credentials(RAW_CREDENTIALS_KEY);
        inOrder.verify(repository, times(1)).save(user);
    }
}

File: src/test/kotlin/cube8540/oauth/authentication/user/application/DefaultUserCredentialsServiceTest.kt
package cube8540.oauth.authentication.user.application

import cube8540.oauth.authentication.users.application.DefaultUserCredentialsService
import cube8540.oauth.authentication.users.domain.*
import io.mockk.every
import io.mockk.mockk
import io.mockk.verifyOrder
import java.util.Optional
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.catchThrowable
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test

class DefaultUserCredentialsServiceTest {

    private val repository: UserRepository = mockk(relaxed = true)
    private val keyGenerator: UserCredentialsKeyGenerator = mockk()
    private val service = DefaultUserCredentialsService(repository)

    init {
        service.keyGenerator = keyGenerator
        every { repository.save(any()) } returnsArgument 0
    }

    @Nested
    inner class GrantCredentialsKeyTest {

        @Test
        fun `request user is not registered in repository`() {
            every { repository.findById(Username("username")) } returns Optional.empty()

            val thrown = catchThrowable { service.grantCredentialsKey("username") }
            assertThat(thrown).isInstanceOf(UserNotFoundException::class.java)
        }

        @Test
        fun `request user is registered in repository`() {
            val user: User = mockk(relaxed = true)

            every { repository.findById(Username("username")) } returns Optional.of(user)

            service.grantCredentialsKey("username")
            verifyOrder {
                user.generateCredentialsKey(keyGenerator)
                repository.save(user)
            }
        }
    }

    @Nested
    inner class CredentialsTest {

        @Test
        fun `request user is not registered in repository`() {
            every { repository.findById(Username("username")) } returns Optional.empty()

            val thrown = catchThrowable { service.accountCredentials("username", "key") }
            assertThat(thrown).isInstanceOf(UserNotFoundException::class.java)
        }

        @Test
        fun `request user is registered in repository`() {
            val user: User = mockk(relaxed = true)

            every { repository.findById(Username("username")) } returns Optional.of(user)

            service.accountCredentials("username", "key")
            verifyOrder {
                user.credentials("key")
                repository.save(user)
            }
        }
    }
}